### PR TITLE
WIP: ENH: Matmul as ufunc

### PR DESCRIPTION
**Pull request by mattip**
The discussion of issue #9028 yielded two different possible directions for allowing `ndarray` subclasses to override `__matmul__` via `__array_ufunc__`. This is option 1 - allow a single `ufunc` `matmul` which will implement vector-vector, vector-matrix, matrix-vector, and matrix-matrix multiplication. In this PR, which builds on ~~PR #11132~~ PR #11175, the implementation of `matmul` is converted to a `ufunc`.

Note the pull request incorporates the changes from pr #11132 so the changesets are repeated

What changed?
- `matmul` is now implemented as a true `ufunc`, it appears in the `umath`'s `n_ops`, can be set and gotten using `PyArray_{G,S}etNumericOps`, and its loops are defined in `loops.c.src`
- tests were expanded to reflect this change

What is not finished?
- ~~since I simply copied the partial loop implementation from `_umath_tests`, `matmul` cannot accept complex or half-float dtypes~~
- ~~I wrote the code together with PR #11132 I probably have chages there that belong here and visa-versa~~
- reuse the logic in `dot` to make this faster

What is next?
- ~~Get feedback on the relative merits of this approach vs. the one taken in PR #11061, where I produced a demi-`ufunc` with a wrapper function that can be used to convert any function into a `ufunc`.~~ The choice has been made, this approach wins.

Edit: update status

### CODE REVIEW

**Comment by mhvk** [comment]
Setup looks great; the trick will indeed be to have a nice way to choose BLAS etc for the inner loop.

**Comment by mhvk** [comment]
@mattip - I think the gufunc loops need some special-case speed-ups that the ufuncs already have, in particular for the case that there is no outer looping to be done, in which case the elementary function should just be called directly. But that is independent of `matmul`...

One thing to possibly consider for `matmul` - the loop selection could include whether or not to use the BLAS routines, or that choice could be made inside the loop function. The latter may be easier to programme, but the former might be better.

**Comment by mattip** [comment]
One of the trickier parts of this is what to do for non-contiguous memory. Just changing the benchmark code from `func(a, a)` to `func(b[::2], [b::2])` where a is 3x3 and b is 6x6 the times for dot and einsum doubled but the time for "dumb" matmul did not change. Increasing the size to 50x50, still with the `[::2]` indexing, now `dot` is 3 times faster than matmul, and dot is over 4 times faster than einsum

**Comment by mhvk** [comment]
Indeed, I think that when the strides are not contiguous, one probably should copy data over, but there is a question for what number of points this becomes useful. My guess would be that for small arrays the copy is so fast that it might not matter.

Aside: I've been implementing this in astropy - this includes a very basic 3x3 matrix multiplication routine, and in my implementation I check in the inner loop whether copies are necessary [1]. As I wrote above, I think it would be more elegant to do this outside, and select a copying or non-copying loop -- and pass on some temporary storage to the inner loop if needed -- though it may not matter that much speed-wise. Also, I think almost anything will be faster than the current call to `einsum`...

[1] https://gist.github.com/mhvk/a3c17f0db77b91151d50a0d278621d97#file-ufunc-c-L363

**Comment by mhvk** [comment]
I like the last commit - but isn't that independent of `matmul`? I.e., should it be build on top of #11175? It would go with the version change...

**Comment by mattip** [comment]
it seems the matmul vector-vector case for column-row -> 2d array was not tested. I put a marker in each case of all the matmul loops, and that one came up missing.

**Comment by mattip** [comment]
note this is a wip, builds on top of master and pr #11175. Once that is merged I will clean up this branch

**Comment by mattip** [comment]
level 1 `blas_{s,d,c,z}axy` can work with non-contiguous data but require zeroing out `y` before the call, does it give a speedup vs. writing the operation `y = a * x` in naive C? I chose to go with the C loop, benchmarking needed.

**Comment by mattip** [comment]
Just noticed there is an optimization for a @ a.T in #6932 and a request for a@ a.H in #6951. Make sure these are not ruined by this refactoring

**Review by tylerjereddy — COMMENTED** [review]
Please rebase on top of the latest master branch so that `.codecov.yml' setting disabling auto-comments can be detected in the PR.

**Comment by jakirkham** [comment]
> Just noticed there is an optimization for a @ a.T in #6932 and a request for a@ a.H in #6951. Make sure these are not ruined by this refactoring

Thanks for taking these into account. :)

We got `a @ a.T` to work before by checking whether the data was the same and only striding changed. `a @ a.H` was a bit hard for us to figure out before given how conjugating will generate a copy.

File: numpy/core/src/umath/_umath_tests.c.src [hunk 1]
```diff
@@ -195,6 +197,41 @@ static void
...
+*/
+
+/*
+ *  This implements the cross product:
+ *        out[n, 0] = in1[n, 1]*in2[n, 2] - in1[n, 2]*in2[n, 1]
+ *        out[n, 1] = in1[n, 2]*in2[n, 0] - in1[n, 0]*in2[n, 2]
+ *        out[n, 2] = in1[n, 0]*in2[n, 1] - in1[n, 1]*in2[n, 0]
+ */
+static void
+@TYPE@_cross1d(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
```
Comment (tylerjereddy) [inline numpy/core/src/umath/_umath_tests.c.src]: I may as well just ask--do we have some docs that describe the mechanics of the `@` symbol usage in NumPy C source? It *appears* to be some kind of mechanism for generalization / substitution, but unless I missed a chapter in the C text book, not standard?

File: numpy/core/src/umath/loops.c.src [hunk 2]
```diff
@@ -2925,6 +2927,501 @@ OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED
...
+            ip1 += chunk * is1;
+            ip2 += chunk * is2;
+            n -= chunk;
+        }
+        *((@type@ *)op) = (@type@)sum;
+    }
+    else
+#endif
+    {
+        @type@ sum = (@type@)0;  /* could make this double */
```
Comment (tylerjereddy) [inline numpy/core/src/umath/loops.c.src]: The comment here means that there's no compelling advantage to having `@type@` of `sum` be flexible, or something else?

File: numpy/core/tests/test_ufunc.py [hunk 3]
```diff
@@ -866,6 +920,89 @@ def test_innerwt_empty(self):
...
+        assert_array_equal(umt.cross1d(a, a), np.zeros((3, 3)))
+        out = np.zeros((3, 3))
+        result = umt.cross1d(a[0], a, out)
+        assert_(result is out)
+        assert_array_equal(result, np.vstack((np.zeros(3), a[2], -a[1])))
+        assert_raises(ValueError, umt.cross1d, np.eye(4), np.eye(4))
+        assert_raises(ValueError, umt.cross1d, a, np.arange(4.))
+        assert_raises(ValueError, umt.cross1d, a, np.arange(3.), np.zeros((3, 4)))
+
+    def test_can_ignore_signature(self):
```
Comment (tylerjereddy) [inline numpy/core/tests/test_ufunc.py]: Any way to split this 70-line "unit test" up into smaller pieces logically?

File: doc/source/reference/c-api.generalized-ufuncs.rst [hunk 4]
```diff
@@ -139,22 +140,35 @@ Notes:
    Each dimension name typically corresponds to one level of looping in the
    elementary function's implementation.
 #. White spaces are ignored.
+#. An integer as a dimension name freezes that dimension to the value,
+#. The name can be suffixed with a question mark, this make the dimension a
```
[the remark this answers is not in the document]
Comment (mattip) [reply doc/source/reference/c-api.generalized-ufuncs.rst]: fixed

File: numpy/core/src/umath/loops.c.src [hunk 5]
```diff
@@ -2925,6 +2927,501 @@ OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED
...
+                /* vector @ matrix, switch ip1, ip2, p and m */
+                @TYPE@_gemv((void*)ip2, is2_n, (void*)ip1, is1_n,
+                            (void*)op, dn, dp);
+            } else if (dp == 1) {
+                /* matrix @ vector */
+                @TYPE@_gemv((void*)ip1, is1_n, (void*)ip2, is2_n,
+                            (void*)op, dm, dn);
+            } else {
+                /* cannot reach here */
+                /* printf("inconceivable to reach this line m=%ld, n=%ld, p=%ld\n",
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/core/src/umath/loops.c.src]: We cannot error inside a loop. I could just remove it, but it is helpful for reasoning about the code and can be enabled for debugging

File: numpy/core/src/umath/loops.c.src [hunk 6]
```diff
@@ -2925,6 +2927,501 @@ OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED
...
+        END_OUTER_LOOP
+        return;
+    }
+    BEGIN_OUTER_LOOP_3
+        char *ip1=args[0], *ip2=args[1], *op=args[2];
+#if @USEBLAS@ & defined(HAVE_CBLAS)
+        /* npy_intp ib1_m = is1_m * dm;
+        npy_intp ob_m = os_m * dp; */
+        /*
+         * TODO: refactor this out to a inner_loop_selector, instead of
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/core/src/umath/loops.c.src]: Maybe, but not in this iteration of the code

File: numpy/core/tests/test_multiarray.py [hunk 7]
```diff
@@ -2706,6 +2711,22 @@ class Sub(np.ndarray):
...
+    def test_matmul_out(self):
+        # overlapping memory
+        a = np.arange(18).reshape(2, 3, 3)
+        b = np.matmul(a, a)
+        c = np.matmul(a, a, out=a)
+        assert_(c is a)
+        assert_equal(c, b)
+        a = np.arange(18).reshape(2, 3, 3)
+        c = np.matmul(a, a, out=a[::-1, ...])
+        assert(c.base is a.base)
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/core/tests/test_multiarray.py]: fixed

File: numpy/core/tests/test_ufunc.py (hunk 3 above)
Comment (mattip) [reply numpy/core/tests/test_ufunc.py]: This come from PR #11175 which must be merged before this one. @mhvk any thoughts?

Comment (mhvk) [reply numpy/core/tests/test_ufunc.py]: Yes, my fault. Might be good to refactor, but out of scope for this PR, I think.

File: numpy/core/src/umath/_umath_tests.c.src (hunk 1 above)
Comment (tylerjereddy) [reply numpy/core/src/umath/_umath_tests.c.src]: Yup, I know that now, but it is not well exposed documentation -- I've been working on #11689 & to be fair the distutils docs [template section](https://github.com/numpy/numpy/blob/master/doc/DISTUTILS.rst.txt#template-files) is nothing more than a placeholder.

**Comment by mattip** [comment]
Closing this, will reopen as a new, cleaner PR

**Pull request closed without merging**